plugin/nfs/nfs.py

```diff
@@ -0,0 +1,421 @@
+                total_size = st.f_blocks * st.f_frsize
+                avail_size = st.f_bavail * st.f_frsize
+                fsid = get_fsid(p)
+                fss.append(
+                    FileSystem(fsid, p, total_size,
+                               avail_size, fsid, self._SYSID))
+            except OSError:
+                pass
+
+        return search_property(fss, search_key, search_value)
+
+    def exports(self, search_key=None, search_value=None, flags=0):
+        """List existing exports"""
+        exports = []
+
+        if os.path.exists(NFSPlugin._EXPORTS):
+            with open(NFSPlugin._EXPORTS, "r") as efile:
+                for line in efile:
+                    ret = NFSPlugin._parse_export(shlex.split(line, '#'))
+                    if ret is None:
+                        continue
+                    exports.append(NFSPlugin._make_object(ret[0], ret[1],
+                                                          ret[2]))
+            efile.close()
+
+        return search_property(exports, search_key, search_value)
+
+    def export_fs(self, fs_id, export_path, root_list, rw_list, ro_list,
+                  anon_uid=NfsExport.ANON_UID_GID_NA,
+                  anon_gid=NfsExport.ANON_UID_GID_NA,
+                  auth_type=None, options=None, flags=None):
+        """Add an export"""
+
+        if fs_id is None and export_path is None:
+            raise LsmError(ErrorNumber.INVALID_ARGUMENT,
+                           'Must provide fs_id or export_path')
+
+        common_opts = []
+        if options is not None:
+            common_opts.extend(options.split(','))
+
+        if anon_uid is not None and anon_uid != NfsExport.ANON_UID_GID_NA:
+            common_opts.append('anonuid=%d' % anon_uid)
+
+        if anon_gid is not None and anon_gid != NfsExport.ANON_UID_GID_NA:
+            common_opts.append('anongid=%d' % anon_gid)
+
+        if auth_type is not None:
+            common_opts.append('sec=%s' % str(auth_type))
+
+        if export_path is None:
+            export_path = NFSPlugin._get_fsid_path(fs_id)
+            if export_path is None:
+                raise LsmError(ErrorNumber.INVALID_ARGUMENT,
+                               'Could not locate filesystem')
+
+        hostlist = rw_list + ro_list
+        for host in rw_list:
+            opts = copy.copy(common_opts)
+            if host in root_list:
+                opts.append('no_root_squash')
+            opts.append('rw')
+            try:
+                NFSPlugin._add_export(export_path, host, opts)
+            except Exception:
+                raise
+
+        for host in ro_list:
+            opts = copy.copy(common_opts)
+            if host in root_list:
+                opts.append('no_root_squash')
+            opts.append('ro')
+            try:
+                NFSPlugin._add_export(export_path, host, opts)
+            except Exception:
+                raise
+
+        NFSPlugin._update_exports()
```

Comment:
> The id of returned NfsExport is None. 
Only if the request to add exports results in creating more than one export entry, there is no way to return multiple exportid's and changing that means fixing all internal and external code that relies on it. To avoid the consequences it instead returns None, which lsmcli at least doesnt mind.


> And lsmcli list --type exports does not merge multiple exports into one.
> You might need to work on export_remove() also for this merging multiple exports.
Can't be done, each exportid represents one unique pair of host+path, which may or may not have the same options as another, but the unique pairing, and the values passed to the kernel are those unique pairings.   To collapse any of those entries you would need to discard the export id's

> If any failure(like define a incorrect export path) occurred, the /etc/exports.d/libstoragemgmt.exports does not clean up.
I have added some more checks to prevent these occuring

> Please simply ignore the exportpath provides by user and use the return of NFSPlugin._get_fsid_path(fs_id).
No. The filesystem ID code looks for the actual separate filesystems on the system, thus what you are suggesting would result in only being able to export the root of and thus the whole of filesystems, which is massively crippling. I have introduced checks to ensure that if both a fsid and path are specified that the path is within the named filesystem.